- Open and read the txt file

In [1]:
with open("data/chapter2/the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
print("Total number of character:", len(raw_text))
print(raw_text[:99])


Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


- Here we open and read the file.
- The file has 20479 characters

# Converting text to tokens.

In [2]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text) # Using regex to split the word on spaces
print(result)


['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']


- Note that even  space is considered here. Another thing to note is that commas and periods are considered part of the word, and hence have to be separated
- Modify to remove commas and periods.

In [3]:
result = re.split(r'([,.]|\s)', text)
print(result)

['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ',', '', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


- Since there are white spaces after every word, it better to remove them. This will save only the words and the punctuations.
- Remove whitespace

In [4]:
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


- Above is an example of how to handle only commas and periods.
- But we need a way to handle all types of punctuations.

In [5]:
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)', text) # Regex for all types of punctuation.
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


- apply to the entire raw_text. This way the entire raw text is now processed.
- This will separate the words and punctuations as separate tokens and will also remove the which spaces.


In [6]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [item for item in preprocessed if item.strip()]
print(len(preprocessed))

4690


- After the processing there are 4690 tokens. Example of the first 30 tokens is printed below.

In [7]:
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


- Since the computer dont function with text, we will give each word/token an ID.
- This way each word will be represented by its ID.
- Below is the logic to assign an ID to each token.
    - First the words are sorted (alphanumerically)
    - After sorting each word is given a number based on its order.

In [8]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


- We can see that there are 1130 distinct words/tokens in our vocabulary.
- So we will create a mapping dictionary to convert them to numbers.
- The dictionary will have 1130 items starting from 0 to 1129.


In [9]:
vocab = {token:integer for integer, token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i>50:
        break

('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)
('His', 51)


# Implement a simple text tokenizer
- It should convert tokens to integers and back.
- Basically, it stores the token -> int and int -> token mappings.
- It is used to encode and decode text.

In [10]:
class SimpleTokenizerV1:
    '''
    A simple class to tokenize text.
    Input is the dictionary of {str: int}
    Creates the reverese mapping -> {int: str}
    Used for encoding string to integer input and decoding the integer back to string.
    '''
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [item for item in preprocessed if item.strip()]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        text = ' '.join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

# Example

In [11]:
tokenizer = SimpleTokenizerV1(vocab)

text = """"It's the last he painted, you know," Mrs. Gisburn said with pardonable pride."""
ids = tokenizer.encode(text)
print(ids)
decoded_text = tokenizer.decode(ids)
print(decoded_text)
# Below is an example of the how the text looks after encoding/converting to its corresponding numbers.

[1, 56, 2, 850, 988, 602, 533, 746, 5, 1126, 596, 5, 1, 67, 7, 38, 851, 1108, 754, 793, 7]
" It' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.


In [12]:
# Let's apply it to some new text
text = "Hello, do you like tea?"
print(tokenizer.encode(text))
# If a word/token is not present in the dictionary, then it will fail

KeyError: 'Hello'

Basically, it is important to have a very large and diverse corpus, so that more words can be in the vocabulary.

# Lets modify the Tokenizer to include unknown words.
- Always have a way to incorporate new words into the vocabulary.

In [13]:
all_tokens = sorted(list(set(preprocessed)))
all_tokens.extend(["<|endoftext|>", "<|unk|>"])
vocab = {token: integer for integer, token in enumerate(all_tokens)}
print(len(vocab.items()))
# added 2 new tokens. One is for end_of_text and the other for unknown.

1132


In [14]:
for i, item in enumerate(list(vocab.items())[-5:]):
    print(item)

('younger', 1127)
('your', 1128)
('yourself', 1129)
('<|endoftext|>', 1130)
('<|unk|>', 1131)


In [15]:
class SimpleTokenizerV2:
    '''
    This is similar to class Simpletokenizerv1 except that it can also handle unknown tokens.
    '''
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        preprocessed = [item if item in self.str_to_int else "<|unk|>" for item in preprocessed]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        text = ' '.join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

# Compare V1 and V2
- The class V1 is able to handle new words that are not in the dictionary.

In [16]:
text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."
text = " <|endoftext|> ".join((text1, text2))
print(text)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the palace.


In [17]:
tokenizer = SimpleTokenizerV2(vocab)
print(tokenizer.encode(text))

[1131, 5, 355, 1126, 628, 975, 10, 1130, 55, 988, 956, 984, 722, 988, 1131, 7]


In [18]:
print(tokenizer.decode(tokenizer.encode(text)))

<|unk|>, do you like tea? <|endoftext|> In the sunlit terraces of the <|unk|>.


- The reason for <|unk|> above is because the original piece of text does not contain the words `Hello` and `palace`.

# Byte Pair Encoding

In [19]:
from importlib.metadata import version
import tiktoken
print(f"tiktoken version: {version('tiktoken')}")

tiktoken version: 0.8.0


In [20]:
tokenizer = tiktoken.get_encoding("gpt2")

In [21]:
text = "Hello, do you like tea? <|endoftext|> In the sunlit terraces of the someunknownPlace."
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 286, 262, 617, 34680, 27271, 13]


In [22]:
# Below is the break down of word someunknownPlace. Note that the space before the word `some` is considered part of the work (` some`)
print(tokenizer.encode(" someunknownPlace"))
print(tokenizer.encode(" some"))
print(tokenizer.encode("unknown"))
print(tokenizer.encode("Place"))


[617, 34680, 27271]
[617]
[34680]
[27271]


In [23]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the someunknownPlace.


In [24]:
text = "Akwirw ier"
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[33901, 86, 343, 86, 220, 959]
Akwirw ier


# Create train and test

In [25]:
# open and encode the text with tiktoken
with open("data/chapter2/the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [26]:
enc_sample = enc_text[50:]
print(enc_sample)

[290, 4920, 2241, 287, 257, 4489, 64, 319, 262, 34686, 41976, 13, 357, 10915, 314, 2138, 1807, 340, 561, 423, 587, 10598, 393, 28537, 2014, 198, 198, 1, 464, 6001, 286, 465, 13476, 1, 438, 5562, 373, 644, 262, 1466, 1444, 340, 13, 314, 460, 3285, 9074, 13, 46606, 536, 5469, 438, 14363, 938, 4842, 1650, 353, 438, 2934, 489, 3255, 465, 48422, 540, 450, 67, 3299, 13, 366, 5189, 1781, 340, 338, 1016, 284, 3758, 262, 1988, 286, 616, 4286, 705, 1014, 510, 26, 475, 314, 836, 470, 892, 286, 326, 11, 1770, 13, 8759, 2763, 438, 1169, 2994, 284, 943, 17034, 318, 477, 314, 892, 286, 526, 383, 1573, 11, 319, 9074, 13, 536, 5469, 338, 11914, 11, 33096, 663, 4808, 3808, 62, 355, 996, 484, 547, 12548, 287, 281, 13079, 410, 12523, 286, 22353, 13, 843, 340, 373, 407, 691, 262, 9074, 13, 536, 48819, 508, 25722, 276, 13, 11161, 407, 262, 40123, 18113, 544, 9325, 701, 11, 379, 262, 938, 402, 1617, 261, 12917, 905, 11, 5025, 502, 878, 402, 271, 10899, 338, 366, 31640, 12, 67, 20811, 1, 284, 910, 11, 351, 10

In [47]:
context_size = 4
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287]
y:      [4920, 2241, 287, 257]


In [48]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, '------>', desired)

[290] ------> 4920
[290, 4920] ------> 2241
[290, 4920, 2241] ------> 287
[290, 4920, 2241, 287] ------> 257


In [50]:
# repeat but with actual words

In [52]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), '------>', tokenizer.decode([desired]))

 and ------>  established
 and established ------>  himself
 and established himself ------>  in
 and established himself in ------>  a


In [53]:
# Efficient data loading

In [54]:
import torch
from torch.utils.data import Dataset, DataLoader

In [55]:
class GPTDatasetV1(Dataset):
    def __init__(self, txt, tokenizer, max_length, stride):
        self.input_ids = []
        self.target_ids = []

        token_ids = tokenizer.encode(txt)

        for i in range(0, len(token_ids)-max_length, stride):
            input_chunk = token_ids[i:i+max_length]
            target_chunk = token_ids[i+1: i+1+max_length]
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))

    def __len__(self):
        return len(self.input_ids)
    
    def __getitem__(self, idx):
        return self.input_ids[idx], self.target_ids[idx]

In [68]:
def create_dataloader_v1(txt, batch_size=4, max_length=256, stride=128, shuffle=True, drop_last=True, num_workers=0):
    tokenizer = tiktoken.get_encoding('gpt2')
    dataset = GPTDatasetV1(txt, tokenizer, max_length, stride)
    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        drop_last=drop_last,
        num_workers=num_workers
    )
    return dataloader

In [71]:
dataloader = create_dataloader_v1(
    raw_text, batch_size=1, max_length=4, stride=1, shuffle=False
)

data_iter = iter(dataloader)
first_batch = next(data_iter)
print(first_batch)

[tensor([[  40,  367, 2885, 1464]]), tensor([[ 367, 2885, 1464, 1807]])]


- The `first_batch` contains tensors for the input_ids and target_ids.
- They are of size 4 since we set the max_length to that value.
- The meaning of stride = 1 can be seen below by fetching the next batch. Basically, the next batch is shifted by 1.

In [72]:
second_batch = next(data_iter)
print(second_batch)

[tensor([[ 367, 2885, 1464, 1807]]), tensor([[2885, 1464, 1807, 3619]])]


# another example

In [73]:
dataloader = create_dataloader_v1(raw_text, batch_size=8, max_length=4, stride=4)

data_iter = iter(dataloader)
inputs, targets = next(data_iter)

print("\nInputs:\n", inputs)
print("\nTargets:\n", targets)


Inputs:
 tensor([[  314,  3114,   379,   262],
        [  287,   262,  6001,   286],
        [  843,   340,   373,   407],
        [  351,   534,   670,  2474],
        [  287,   262, 13203,  5482],
        [  318,    11,   355,   257],
        [14482,  1625,   503,   438],
        [  319,   326,   966,   314]])

Targets:
 tensor([[ 3114,   379,   262, 50085],
        [  262,  6001,   286,   465],
        [  340,   373,   407,   691],
        [  534,   670,  2474,   198],
        [  262, 13203,  5482,  1044],
        [   11,   355,   257,  3896],
        [ 1625,   503,   438,   439],
        [  326,   966,   314,   714]])


- Note that the stride is 4. 
- This will utilize the entire dataset and will not skip a single word.
- It will also avoid any overlap between batches to prevent overfitting.

In [74]:
vocab_size = 50257
output_dim = 256
token_embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

In [75]:
token_embeddings = token_embedding_layer(inputs)
print(token_embeddings)

tensor([[[-1.1851, -0.6410, -0.0301,  ..., -0.4475,  1.3675, -0.7273],
         [-1.9013,  0.4860, -0.2807,  ...,  0.0481, -0.3087,  0.1891],
         [-1.7342,  0.5630, -0.4547,  ...,  1.0605,  0.8760,  0.0233],
         [ 0.3482,  0.7969, -0.6621,  ...,  0.4677,  1.0594, -0.3324]],

        [[-0.1724, -1.1555, -0.2644,  ...,  0.7477, -0.6358,  0.4778],
         [ 0.3482,  0.7969, -0.6621,  ...,  0.4677,  1.0594, -0.3324],
         [-0.4438, -1.4941, -0.1680,  ..., -0.3523, -1.1681,  0.6715],
         [-0.1678,  0.5555, -0.6539,  ..., -0.7922, -0.2425,  0.6161]],

        [[ 0.0346,  0.3546, -0.6308,  ...,  1.5669, -0.5016, -2.6712],
         [-0.0094,  0.4048, -0.7069,  ...,  1.7612,  0.7381, -0.2014],
         [ 0.7656, -0.1843,  1.4468,  ..., -0.3029,  0.3344, -0.8753],
         [-1.3456,  0.2740, -1.3920,  ..., -0.3272, -0.2887, -0.2970]],

        ...,

        [[ 1.4827,  0.1429, -2.0626,  ..., -0.2451, -0.1450, -1.1961],
         [-0.3941,  0.4757,  2.6194,  ..., -0.4097,  1.05

In [76]:
print(token_embeddings.shape)

torch.Size([8, 4, 256])


- For the GPT model's absolute embedding approach, we just need to create another embedding layer that has the same dimension as the token_embedding_layer

In [81]:
max_length = 4
dataloader = create_dataloader_v1(
    raw_text,
    batch_size=8,
    max_length=max_length,
    stride=max_length,
    shuffle=False
)
data_iter = iter(dataloader)
inputs, targets = next(data_iter)
print(inputs)
print(inputs.shape)

token_embeddings = token_embedding_layer(inputs)
print(token_embeddings.shape)

context_length = max_length
pos_embedding_layer = torch.nn.Embedding(context_length, output_dim)
pos_embeddings = pos_embedding_layer(torch.arange(context_length))
print(pos_embeddings.shape)


tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])
torch.Size([8, 4])
torch.Size([8, 4, 256])
torch.Size([4, 256])


In [82]:
input_embeddings = context_length + pos_embeddings

tensor([0, 1, 2, 3])